### PR TITLE
v3.0.16 hotfix

### PR DESCRIPTION
**Pull request by imtbkcat**
add log to print SQL and stack when field length is set to `MaxDecimalWidth`

### CODE REVIEW

File: expression/builtin_arithmetic.go [hunk 1]
```diff
@@ -136,11 +141,13 @@ func (c *arithmeticDivideFunctionClass) setType4DivDecimal(retTp, a, b *types.Fi
 	}
 	if a.Flen == types.UnspecifiedLength {
 		retTp.Flen = mysql.MaxDecimalWidth
+		logutil.Logger(context.Background()).Warn("set field type Flen to MaxDecimalWidth", zap.Error(ErrUnknown.GenWithStackByArgs()))
```
Comment (bb7133) [inline expression/builtin_arithmetic.go:144]: Please try to get the SQL statement from `ctx` by either passing `ctx` to `setType4DivDecimal`, or checking the `Flen` from the caller(https://github.com/pingcap/tidb/blob/b4644d433a8231de28c4e25bcc76c7c81e44c427/expression/builtin_arithmetic.go#L619)

Either way should be trivial enough.

File: expression/builtin_arithmetic.go [hunk 2]
```diff
@@ -136,11 +141,13 @@ func (c *arithmeticDivideFunctionClass) setType4DivDecimal(retTp, a, b *types.Fi
...
 		return
 	}
 	retTp.Flen = a.Flen + decb + precIncrement
 	if retTp.Flen > mysql.MaxDecimalWidth {
 		retTp.Flen = mysql.MaxDecimalWidth
+		logutil.Logger(context.Background()).Warn("set field type Flen to MaxDecimalWidth", zap.Error(ErrUnknown.GenWithStackByArgs()))
```
Comment (bb7133) [inline expression/builtin_arithmetic.go:150]: Ditto

File: expression/aggregation/base_func.go [hunk 3]
```diff
@@ -155,6 +160,7 @@ func (a *baseFuncDesc) typeInfer4Avg(ctx sessionctx.Context) {
 			a.RetTp.Decimal = mathutil.Min(a.Args[0].GetType().Decimal+types.DivFracIncr, mysql.MaxDecimalScale)
 		}
 		a.RetTp.Flen = mysql.MaxDecimalWidth
+		logutil.Logger(context.Background()).Warn("set field type Flen to MaxDecimalWidth", zap.String("sql", ctx.GetSessionVars().StmtCtx.OriginalSQL), zap.Error(expression.ErrUnknown.GenWithStackByArgs()))
```
Comment (bb7133) [inline expression/aggregation/base_func.go:163]: > zap.Error(expression.ErrUnknown.GenWithStackByArgs())

The stack cannot be printed in this way, you can check the log here by a simple statement:
```
select sum(a) from t;
```
whereas `a` is of type `DECIMAL`

I found that using `zap.Stack("stack")` just works.

File: expression/builtin_arithmetic.go (hunk 1 above)
Comment (imtbkcat) [reply expression/builtin_arithmetic.go:144]: I add log in the caller function.

File: expression/builtin_arithmetic.go [hunk 4]
```diff
@@ -592,6 +617,7 @@ func (c *arithmeticDivideFunctionClass) getFunction(ctx sessionctx.Context, args
 	}
 	bf := newBaseBuiltinFuncWithTp(ctx, args, types.ETDecimal, types.ETDecimal, types.ETDecimal)
 	c.setType4DivDecimal(bf.tp, lhsTp, rhsTp)
+	logutil.Logger(context.Background()).Warn("set field type Flen to MaxDecimalWidth", zap.String("sql", ctx.GetSessionVars().StmtCtx.OriginalSQL), zap.Stack("stack"))
```
Comment (bb7133) [inline expression/builtin_arithmetic.go:620]:
```
if bf.tp.Flen == mysql.MaxDecimalWidth {
			logutil.Logger(context.Background()).Warn("set field type Flen to MaxDecimalWidth", zap.String("sql", ctx.GetSessionVars().StmtCtx.OriginalSQL), zap.Stack("stack"))
```
How about this? `bf.tp` may not set to max decimal length here.

File: expression/builtin_arithmetic.go [hunk 5]
```diff
@@ -821,6 +847,7 @@ func (c *arithmeticModFunctionClass) setType4ModRealOrDecimal(retTp, a, b *types
 		retTp.Flen = mathutil.Max(a.Flen, b.Flen)
 		if isDecimal {
 			retTp.Flen = mathutil.Min(retTp.Flen, mysql.MaxDecimalWidth)
+			logutil.Logger(context.Background()).Warn("set field type Flen to MaxDecimalWidth", zap.Error(ErrUnknown.GenWithStackByArgs()))
```
Comment (bb7133) [inline expression/builtin_arithmetic.go:850]: Ditto

**Pull request closed without merging**